crates/synthesis/src/qsd.rs

```diff
@@ -0,0 +1,800 @@
+        None,
+        VWType::OnlyW,
+    )?;
+    let (right_circuit, _, wmat_a) =
+        demultiplex(&a1, &a2, opt_a1_val, opt_a2_val, depth, None, VWType::OnlyV)?;
+
+    // middle circ
+    // zmat is needed in order to reduce two cz gates, and combine them into the B2 matrix
+    let mut zmat = DMatrix::<Complex64>::zeros(dim / 2, dim / 2);
+    for i in 0..dim / 2 {
+        zmat[(i, i)] = if i < dim / 4 {
+            Complex64::from(1.0)
+        } else {
+            Complex64::from(-1.0)
+        };
+    }
+    // wmatA and vmatC are combined into B1 and B2
+    let b1 = &wmat_a * &vmat_c;
+    let b2 = if opt_a1_val {
+        &zmat * &wmat_a * &b * &vmat_c * &zmat
+    } else {
+        &wmat_a * &b * &vmat_c
+    };
+    let middle_circ = demultiplex(&b1, &b2, opt_a1_val, opt_a2_val, depth, None, VWType::All)?.0;
+
+    // the output circuit of the block ZXZ decomposition from [2]
+    let qr = (0..num_qubits).map(Qubit::new).collect::<Vec<_>>();
+    append(&mut out, left_circuit, &qr)?;
+    let _ = out.push_standard_gate(StandardGate::H, &[], &[Qubit((num_qubits - 1) as u32)]);
+    append(&mut out, middle_circ, &qr)?;
+    let _ = out.push_standard_gate(StandardGate::H, &[], &[Qubit((num_qubits - 1) as u32)]);
+    append(&mut out, right_circuit, &qr)?;
+    if opt_a2_val && depth == 0 && dim > 4 {
+        apply_a2(&out, two_qubit_decomposer)
+    } else {
+        Ok(out)
+    }
+}
+
+fn _zxz_decomp_svd(a: DMatrix<Complex64>) -> (DMatrix<Complex64>, DMatrix<Complex64>) {
+    let (v, sigma, w_dg) = svd_decomposition(&a);
+
+    let s = &v * &sigma * &v.adjoint();
+    let u = v * w_dg;
+
+    (s, u)
+}
+
+/// Block ZXZ decomposition method, by Krol and Al-Ars [2]
+fn _block_zxz_decomp(
+    mat: &DMatrix<Complex64>,
+) -> (
+    DMatrix<Complex64>,
+    DMatrix<Complex64>,
+    DMatrix<Complex64>,
+    DMatrix<Complex64>,
+) {
+    debug_assert!(verify_unitary(mat));
+
+    let i = Complex64::new(0.0, 1.0);
+    let n = mat.shape().0 / 2;
+    let x = mat.view((0, 0), (n, n));
+    let y = mat.view((0, n), (n, n));
+    let u21 = mat.view((n, 0), (n, n));
+    let u22 = mat.view((n, n), (n, n));
+    let (sx, ux) = _zxz_decomp_svd(x.into_owned());
+    let (sy, uy) = _zxz_decomp_svd(y.into_owned());
+    let c = ((&uy.adjoint() * &ux) * i).adjoint();
+    let a1 = (sx + sy * i) * &ux;
+    let a2 = u21 + (u22 * (uy.adjoint() * ux) * i);
+    let b = (a1.adjoint() * x) * Complex64::from(2.0) - DMatrix::<Complex64>::identity(n, n);
+    (a1, a2, b, c)
+}
+
+/// Verify ZXZ decomposition gives the same unitary
+#[allow(clippy::toplevel_ref_arg)]
+fn _zxz_decomp_verify(
+    mat: &DMatrix<Complex64>,
+    a1: &DMatrix<Complex64>,
+    a2: &DMatrix<Complex64>,
+    b: &DMatrix<Complex64>,
+    c: &DMatrix<Complex64>,
+) -> bool {
+    let n = mat.shape().0 / 2;
+    let iden = DMatrix::<Complex64>::identity(n, n);
+
+    let a_block = stack![a1, 0; 0, a2];
+
+    let b1 = iden.clone() + b;
+    let b2 = iden.clone() - b;
```

Comment:
This code will only be used as a part of self-sanity check in the non-release compilation but anyhow the cloning here can be removed:
```suggestion
    let b1 = &iden + b;
    let b2 = &iden - b;
```